### PR TITLE
Fix token inflation

### PR DESCRIPTION
**Pull request by guilherme-brandao**
Readded the halving mechanism mimicking BTC tokenomics but considering these parameters:

```
blocksPerYear      = 6311520 // ~5s block
totalALLO             = 1000000000 // 1 billion
```

Removed the `InitChainer` command because I saw that we need to add `inflation` and `annual_provisions` on the `genesis.json`.

jsyk the inflation calculated was ~15% and the annual provisions 15,027,729,120 tokens.

I will be making a PR to add it on the networks repo.

The mint module on the genesis will be:
```
"mint": {
      "minter": {
        "inflation": "0.15030000000000000",
        "annual_provisions": "15027729120.0"
      },
      "params": {
        "mint_denom": "uallo", 
        "inflation_rate_change": "0.150300000000000000",
        "inflation_max": "0.150300000000000000",
        "inflation_min": "0.000000000000000000",
        "goal_bonded": "0.670000000000000000",
        "blocks_per_year": "6311520"
      }
    },
```
_________________________________ * NEW *___________________________________________

After understanding how the Cosmos SDK's mint module worked, I opted to fork the mint module and customize it to achieve the goal of this PR.

The goal is to implement halvings in the token's inflation effectively and a total max supply.

The main files to look at in this PR are:

`module_test.go` -> test triggering the `BeginBlocker`
`abci.go` -> Implementation of `BeginBlocker`, where new inflation will be calculated according to the `BlockHeight`
`minter_test.go` -> test of the minter functions
`minter.go` -> Implementation of the code that effectively does the halving and calculates the new annual provisions.

The other modifications revolved around standardizing the module so that it is in line with our emissions module.

### CODE REVIEW

**Review by kpeluso — APPROVED** [review]
+1 something like what Tyler said but gtg otherwise.

**Comment by okedeji** [comment]
Looking forward to his PR to the network repo. should make sure that the PR is to generate-genesis file and not genesis.json. could help out if help is needed @guilherme-brandao

File: x/mint/module/abci.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	"time"
+
+	cosmosMath "cosmossdk.io/math"
+	"github.com/allora-network/allora-chain/x/mint/keeper"
+	"github.com/allora-network/allora-chain/x/mint/types"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// TODO: MaxSupply should be defined somewhere else
```
Comment (relyt29) [inline x/mint/module/abci.go:14]: lets put this in a params object that is pulled in and exported/imported from/to the genesis.json file

can reference genesis.go in the emissions module for how I did that for that module https://github.com/allora-network/allora-chain/blob/383c8369ac6ab1b7dda2fb73b4a442ebb4d0d930/x/emissions/genesis.go#L8

File: x/mint/proto/mint/v1beta1/types.proto [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  ];
+  // goal of percent bonded atoms
+  string goal_bonded = 5 [
+    (cosmos_proto.scalar)  = "cosmos.Dec",
+    (gogoproto.customtype) = "cosmossdk.io/math.LegacyDec",
+    (gogoproto.nullable)   = false,
+    (amino.dont_omitempty) = true
+  ];
+  // expected blocks per year
+  uint64 blocks_per_year = 6;
```
Comment (relyt29) [inline x/mint/proto/mint/v1beta1/types.proto:61]: I would put `MaxSupply` from abci.go in here and store it in the keeper to pull in abci.go

**Review by relyt29 — APPROVED** [review]
2 minor comments, pinged you on slack with a question

otherwise, I'm good on this

**Comment by relyt29** [comment]
```
failed to validate genesis state: inflation rate change too large: 357.358262400000000000 [cosmos/cosmos-sdk@v0.50.3/x/genutil/client/cli/gentx.go:96]
````

File: x/mint/types/minter_test.go [hunk 3]
```diff
@@ -1,60 +0,0 @@
...
-
-import (
-	"testing"
-
-	"github.com/stretchr/testify/require"
-
-	"cosmossdk.io/math"
-)
-
-func TestNextInflationRate(t *testing.T) {
```
Comment (relyt29) [inline x/mint/types/minter_test.go]: uhhhhhhhhhhh are we not going to have tests ????

File: x/mint/types/params.go [hunk 4]
```diff
@@ -11,26 +11,30 @@ import (
...
-		BlocksPerYear:       uint64(60 * 60 * 8766 / 5), // assuming 5 second block times
+		MintDenom:             sdk.DefaultBondDenom,
+		InflationRateChange:   math.LegacyNewDecWithPrec(3573582624, 7),
+		InflationMax:          math.LegacyNewDecWithPrec(3573582624, 7),
+		InflationMin:          math.LegacyNewDecWithPrec(0, 2),
+		GoalBonded:            math.LegacyNewDecWithPrec(67, 2),
+		BlocksPerYear:         uint64(60 * 60 * 8766 / 5),     // assuming 5 second block times
+		MaxSupply:             "1000000000000000000000000000", //1 billion allo * 1e18 (exponent) = 1e27 uallo
+		HalvingInterval:       uint64(25246080),
+		CurrentBlockProvision: "2831000000000000000000", // uallo per block
```
Comment (relyt29) [inline x/mint/types/params.go:37]: same with this, could just be a uint directly

File: x/mint/proto/mint/v1beta1/types.proto [hunk 5]
```diff
@@ -59,4 +59,10 @@ message Params {
   ];
   // expected blocks per year
   uint64 blocks_per_year = 6;
+  // maximum total supply of the coin
+  string max_supply  = 7;
```
Comment (relyt29) [inline x/mint/proto/mint/v1beta1/types.proto:63]:
```
string max_supply = 7 [
    (cosmos_proto.scalar)  = "cosmos.Uint",
    (gogoproto.customtype) = "cosmossdk.io/math.Uint",
    (gogoproto.nullable)   = false,
    (amino.dont_omitempty) = true
  ];
  ```

File: x/mint/types/minter.go [hunk 6]
```diff
@@ -19,17 +17,19 @@ func NewMinter(inflation, annualProvisions math.LegacyDec) Minter {
...
 	)
 }
 
 // DefaultInitialMinter returns a default initial Minter object for a new chain
-// which uses an inflation rate of 13%.
+// which uses an inflation rate of 15,03%.
 func DefaultInitialMinter() Minter {
 	return InitialMinter(
-		math.LegacyNewDecWithPrec(13, 2),
+		math.LegacyNewDecWithPrec(3573582624, 7),
```
Comment (relyt29) [inline x/mint/types/minter.go:32]: is this number right lol?

File: x/mint/types/params.go [hunk 7]
```diff
@@ -168,3 +169,16 @@ func validateBlocksPerYear(i interface{}) error {
...
 	return nil
 }
+
+func validateMaxSupply(i interface{}) error {
+	v, ok := i.(string)
+	if !ok {
+		return fmt.Errorf("invalid parameter type: %T", i)
+	}
+
+	if strings.TrimSpace(v) == "" {
```
Comment (relyt29) [inline x/mint/types/params.go:179]: can get rid of this if you make maxSupply a uint not a string

File: x/mint/types/minter_test.go (hunk 3 above)
Comment (guilherme-brandao) [reply x/mint/types/minter_test.go]: logic was moved to `abci.go` and it's being tested in `module_test.go`

File: x/mint/types/minter.go (hunk 6 above)
Comment (guilherme-brandao) [reply x/mint/types/minter.go:32]: It's the calculated initial inflation. Inflation now it's just used for query purposes, it will be updated every block.

**Review by relyt29 — APPROVED** [review]
lgtm merge and :ship: it

**Pull request merged**